src/errors.rs

```diff
@@ -1021,3 +1021,106 @@ pub enum LoadJsonError {
 pub type StoreJsonError = serde_json::Error;
 
 pub type StoreTomlError = toml_edit::ser::Error;
+
+/// Format a short list with commas and an "and" before the last item in a multi-item list.
+///
+/// If there are 2 or fewer items, they are always displayed (regardless of formatting width
+/// limit), and the first item is also always displayed.
+///
+/// The formatting width limit pertains only to the total length of the items in the list, not
+/// additional characters/text added when formatting.
+///
+/// The default width is 40 characters.
+pub struct FormatShortList<S> {
+    items: Vec<S>,
+}
+
+impl<S: AsRef<str>> FormatShortList<S> {
+    pub fn new(mut items: Vec<S>) -> Self {
+        // To keep the display compact, sort by name length and truncate long lists.
+        // We first sort by name because rust defaults to a stable sort and this will
+        // have by-name as the tie breaker.
+        items.sort_by(|a, b| {
+            let a = a.as_ref();
+            let b = b.as_ref();
+            a.len().cmp(&b.len()).then_with(|| a.cmp(b))
+        });
+        FormatShortList { items }
+    }
+
+    pub fn string(items: Vec<S>) -> String {
+        Self::new(items).to_string()
+    }
+}
+
+impl<S: AsRef<str>> Display for FormatShortList<S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        use std::fmt::Write;
+
+        let width = f.width().unwrap_or(40);
+
+        // Below the character count is obtained with `len`, so these must be ASCII.
```

Comment:
I saw that. I didn't think it would matter as there's no escape sequences (and from the perspective of this implementation, it's just string formatting).

Reply:
I think the main thing would be unicode characters. Something like é can be written either as a single code point, or as a combination of multiple unicode code points (`"e\u0301"` vs `"\u00e9"`). The first case would be counted incorrectly for example.

I think we should do the right thing here even if it ends up working out OK.

Reply:
Ugh, I wasn't thinking about unicode modifiers. I'll use `measure_text_width` then since it uses `unicode-width`. And then we can add terminal colors safely :)

Reply:
Though then again I don't think most of the data going through this (like crate names) can have unicode, but better safe than sorry.